src/control/controlvalue.h

```diff
@@ -1,60 +1,67 @@
-#ifndef CONTROLVALUE_H
-#define CONTROLVALUE_H
+#pragma once
 
 #include <limits>
 
 #include <QAtomicInt>
 #include <QObject>
 
-#include "util/compatibility.h"
 #include "util/assert.h"
+#include "util/compatibility.h"
 
 // for lock free access, this value has to be >= the number of value using threads
 // value must be a fraction of an integer
-const int cRingSize = 8;
+constexpr int kDefaultRingSize = 8;
 // there are basically unlimited readers allowed at each ring element
 // but we have to count them so max() is just fine.
 // NOTE(rryan): Wrapping max with parentheses avoids conflict with the max macro
 // defined in windows.h.
-const int cReaderSlotCnt = (std::numeric_limits<int>::max)();
+constexpr int kMaxReaderSlots = (std::numeric_limits<int>::max)();
 
 // A single instance of a value of type T along with an atomic integer which
 // tracks the current number of readers or writers of the slot. The value
-// m_readerSlots starts at cReaderSlotCnt and counts down to 0. If the value is
+// m_readerSlots starts at kMaxReaderSlots and counts down to 0. If the value is
 // 0 or less then reads to the value fail because there are either too many
 // readers or a write is occurring. A write to the value will fail if
-// m_readerSlots is not equal to cReaderSlotCnt (e.g. there is an active
+// m_readerSlots is not equal to kMaxReaderSlots (e.g. there is an active
 // reader).
-template<typename T>
+template <typename T>
 class ControlRingValue {
   public:
     ControlRingValue()
-        : m_value(T()),
-          m_readerSlots(cReaderSlotCnt) {
+      : m_readerSlots(kMaxReaderSlots) {
     }
 
+    // Tries to copy the stored value if a reader slot is available.
+    // This operation can be repeated multiple times for the same
+    // slot, because the stored value is preserved.
     bool tryGet(T* value) const {
         // Read while consuming one readerSlot
-        bool hasSlot = (m_readerSlots.fetchAndAddAcquire(-1) > 0);
-        if (hasSlot) {
+        if (m_readerSlots.fetchAndAddAcquire(-1) > 0) {
             *value = m_value;
+            m_readerSlots.fetchAndAddRelease(1);
+            return true;
+        } else {
+            // Otherwise a writer is active. The writer will reset
+            // the counter in m_readerSlots when releasing the lock
+            // and we must not re-add the substracted value here!
+            return false;
         }
-        (void)m_readerSlots.fetchAndAddRelease(1);
-        return hasSlot;
     }
 
     bool trySet(const T& value) {
         // try to lock this element entirely for reading
-        if (m_readerSlots.testAndSetAcquire(cReaderSlotCnt, 0)) {
+        if (m_readerSlots.testAndSetAcquire(kMaxReaderSlots, 0)) {
             m_value = value;
-            m_readerSlots.fetchAndAddRelease(cReaderSlotCnt);
+            m_readerSlots.fetchAndAddRelease(kMaxReaderSlots);
             return true;
+        } else {
```

Comment:
isn't this else redundant?

Reply:
Why?? Sometimes early-exit-style make sense for error checking, but here here we have an explicit either/or decision.

Reply:
.. because we always return in the if branch above. I do not mind much about this because the "else" will be optimized out, but know that some code analysis tools complains about this.